Source/Scene/ArcGisMapServerImageryProvider.js

```diff
@@ -297,6 +301,91 @@ define([
+            var projected = imageryProvider._tilingScheme.projection.project(new Cartographic(longitude, latitude, 0.0));
+            horizontal = projected.x;
+            vertical = projected.y;
+            sr = '3857';
+        }
+
+        var url = imageryProvider._url + '/identify?f=json&tolerance=2&geometryType=esriGeometryPoint';
+        url += '&geometry=' + horizontal + ',' + vertical;
+        url += '&mapExtent=' + rectangle.west + ',' + rectangle.south + ',' + rectangle.east + ',' + rectangle.north;
+        url += '&imageDisplay=' + imageryProvider._tileWidth + ',' + imageryProvider._tileHeight + ',96';
+        url += '&sr=' + sr;
+
+        url += '&layers=visible';
+        if (defined(imageryProvider._layers)) {
+            url += ':' + imageryProvider._layers;
+        }
+
+        if (defined(imageryProvider._token)) {
+            url += '&token=' + imageryProvider._token;
+        }
+
+        if (defined(imageryProvider._proxy)) {
+            url = imageryProvider._proxy.getURL(url);
+        }
+
+        return url;
+    }
+
+    function jsonToFeatures(json) {
+        var result = [];
+
+        var features = json.results;
+        if (!defined(features)) {
+            return result;
+        }
+
+        for (var i = 0; i < features.length; ++i) {
+            var feature = features[i];
+
+            var featureInfo = new ImageryLayerFeatureInfo();
+            featureInfo.data = feature;
+            featureInfo.name = feature.value;
+            featureInfo.properties = feature.attributes;
+            featureInfo.configureDescriptionFromProperties(feature.attributes);
+
+            // If this is a point feature, use the coordinates of the point.
+            if (feature.geometryType === 'esriGeometryPoint' && feature.geometry) {
+                var wkid = feature.geometry.spatialReference && feature.geometry.spatialReference.wkid ? feature.geometry.spatialReference.wkid : 4326;
+                if (wkid === 4326 || wkid === 4283) {
+                    featureInfo.position = Cartographic.fromDegrees(feature.geometry.x, feature.geometry.y, feature.geometry.z);
+                } else if (wkid === 102100 || wkid === 900913 || wkid === 3857) {
+                    var projection = new WebMercatorProjection();
+                    featureInfo.position = projection.unproject(new Cartesian3(feature.geometry.x, feature.geometry.y, feature.geometry.z));
+                }
+            }
+
+            result.push(featureInfo);
+        }
+
+        return result;
+    }
+
+    function updateToken(imageryProvider) {
+        if (!defined(imageryProvider._newTokenRequestInFlight) && defined(imageryProvider._requestNewToken)) {
+            imageryProvider._newTokenRequestInFlight = imageryProvider._requestNewToken().then(function(newToken) {
+                delete imageryProvider._newTokenRequestInFlight;
+                imageryProvider.token = newToken;
+                return newToken;
+            });
```

Comment:
Super subtle bug possibility here.  What happens if `imageryProvider._requestNewToken()` returns a promise that is already resolved?  In the version of when.js we're currently using, the function passed to `.then` will be executed immediately, before the call to `then()` even returns.  Which means `_newTokenRequestInFlight` will be set to undefined before it even gets assigned the promise in the first place!  And then when `then()` returns, the (already resolved) promise it returns will be assigned to `imageryProvider._newTokenRequestInFlight`.  End result: we think there's a request in flight when there's actually not.

Reply:
You can fix this using an approach like this:
```
var loadPromise = imageryProvider._requestNewToken().then(function(newToken) {
    imageryProvider.token = newToken;
    return newToken;
});

var resolvedOrRejected = false;
loadPromise.always(function() {
    resolvedOrRejected = true;
    imageryProvider._newTokenRequestInFlight = undefined;
});

imageryProvider._newTokenRequestInFlight = resolvedOrRejected ? undefined : loadPromise;
```

Yes, this is kind of terrible, and is one reason that all promise implementations have moved to not immediately calling the callback.

Reply:
Thanks.

8c079608751246ac1a0870cdc2e4e208aca8a5dd